src/Analyzer/Passes/QueryAnalysisPass.cpp

```diff
@@ -2868,12 +2887,56 @@ QueryAnalyzer::QueryTreeNodesWithNames QueryAnalyzer::resolveUnqualifiedMatcher(
     auto table_expressions_stack = buildTableExpressionsStack(nearest_query_scope_query_node->getJoinTree());
     std::vector<QueryTreeNodesWithNames> table_expressions_column_nodes_with_names_stack;
 
+    std::unordered_set<std::string> left_table_expression_column_names_to_skip;
+    std::unordered_set<std::string> right_table_expression_column_names_to_skip;
+
     for (auto & table_expression : table_expressions_stack)
     {
         QueryTreeNodesWithNames matched_expression_nodes_with_column_names;
 
         if (auto * array_join_node = table_expression->as<ArrayJoinNode>())
+        {
+            size_t table_expressions_column_nodes_with_names_stack_size = table_expressions_column_nodes_with_names_stack.size();
+            if (table_expressions_column_nodes_with_names_stack_size < 1)
+                throw Exception(ErrorCodes::LOGICAL_ERROR,
+                    "Expected at least 1 table expressions on stack before ARRAY JOIN processing. Actual {}",
+                    table_expressions_column_nodes_with_names_stack_size);
+
+            auto & table_expression_column_nodes_with_names = table_expressions_column_nodes_with_names_stack.back();
+
+            const auto & array_join_column_list = array_join_node->getJoinExpressions();
+            const auto & array_join_column_nodes = array_join_column_list.getNodes();
+
+            /** Special case with ARRAY JOIN column without alias.
+              *
+              * CREATE TABLE test_table (id UInt64, value String, value_array Array(UInt8)) ENGINE=TinyLog;
+              * SELECT * FROM test_table ARRAY JOIN value_array;
+              *
+              * In matched columns `value_array` must be resolved into array join column.
+              */
+            for (const auto & array_join_column_node : array_join_column_nodes)
+            {
+                if (array_join_column_node->hasAlias())
```

Comment:
Didn't understand that.

Reply:
We can JOIN with ARRAY JOIN columns, and for JOIN ON section we support accessing ARRAY JOIN columns from aliases, but for JOIN with USING section, we try to resolve column USING only from left and right table expression. And to find this column we must add such check here.